glue/plugins/tools/spectrum_tool/qt/spectrum_tool.py

```diff
@@ -93,25 +83,38 @@ def spectrum(data, attribute, roi, slc, zaxis):
             posdata = np.argwhere(slice_data)
 
             # mask for each slice
-            mask = roi.contains(posdata[:, 0], posdata[:, 1])
+            mask = roi.contains(posdata[:, 0], posdata[:, 1]).\
+                reshape((data.shape[xaxis], data.shape[yaxis]))
+
+            # repeat mask along zaxis
+            repeat_shape = [1, 1, 1]
+            repeat_shape[xaxis] = data.shape[xaxis]
+            repeat_shape[yaxis] = data.shape[yaxis]
+            mask = mask.reshape(repeat_shape).repeat(nz, zaxis)
+
+            new_shape = [1, 1, 1]
+            new_shape[zaxis] = nz
+            new_shape[xaxis] = -1
+            data = data[mask].reshape(new_shape)  # data[mask].shape is (n, )
+            finite = np.isfinite(data)
 
-            return_data = []
-            for i in range(nz):
-                select_data = np.take(data, i, axis=zaxis).ravel()[mask]
-                finite = np.isfinite(select_data).sum()
-                return_data.append(1. * np.nansum(select_data) / finite)
+        try:
+            assert data.ndim == ndim
 
-            data = np.array(return_data)
+            for i in reversed(list(range(ndim))):
+                if i != zaxis:  # index of the zaxis in dataset
+                    data = np.nansum(data, axis=i)
+                    finite = finite.sum(axis=i)
```

Comment:
Just to make sure I understand, does this change the behavior of the spectrum extractor for e.g. 4D cubes? (did it work before?)